pkg/endpoint/bpf.go

```diff
@@ -366,7 +433,6 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir string) error {
 	createdIPv4EgressMap := false
 	defer func() {
 		if err != nil {
-			e.Mutex.Lock()
 			if createdPolicyMap {
 				// Remove policy map file only if it was created
 				// in this update cycle
```

Comment:
the e.Consumable on the next line doesn't need to hold the lock()?

Reply:
It does, but I think there is a possible race condition where the consumable WAS non-nil but was changed by another goroutine after we released the lock (i.e., endpoint's identity changed). To correctly implement the intended semantics (RemoveMap only if we added it, and to remove it from the same consumable we added it to) we need keep the original e.Consumable in a local variable and use that instead of e.Consumable at error case cleanup. Since we have a local reference now no locking is needed.

Thanks for pointing this out!

Reply:
Just to be sure, so the consumable itself doesn't need to have its lock held while changing its values?

Reply:
Consumable's AddMap() and RemoveMap() methods take the lock on it, so they should be called with that lock not held.

Setting e.PolicyMap to nil in the error case seems dodgy, however. Endpoint's mutex is needed for that after all.

Thanks for digging deep enough on this.

Reply:
Will squash in a fix for this reverting the removal of locking in this block & ask for one more review right after. Thanks!

Reply:
OK, I squashed in the fix that reverts the removal of e.Mutex locking here. Not sure what I was thinking at the time I removed them, but at least we found a potential issue with the consumable being changed in between.